File: C++/BackTracking/03_word_break.cpp
/*
Idea:
using loop to find word which is in dictionary & recursively calling itself for next words till all words match the dictionary
*/

//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
// User function Template for C++

class Solution
{
    set<string> hash; // Set to store valid dictionary words
    vector<string> ans; // Vector to store valid word break combinations

public:
    // Recursive function to find word break combinations
    void solve(string s, int ind)
    {
        if (ind == s.size())
        {
            s.pop_back(); // Remove the trailing space
            ans.push_back(s); // Store the valid word break combination
            return;
        }

        string subS = "";
        for (int i = ind; i < s.size(); ++i)
        {
            subS += s[i]; // Build a substring
            if (hash.find(subS) != hash.end())
            { // Check if the substring is a valid word
                string temp = s;
                temp.insert(i + 1, " "); // Insert a space after the valid word
                solve(temp, i + 2); // Recur to the next position after the inserted space
            }
        }
    }

    // Function to find all valid word break combinations
    vector<string> wordBreak(int n, vector<string>& dict, string s)
    {
        for (auto it : dict)
            hash.insert(it); // Populate the set with valid dictionary words

        solve(s, 0); // Start the recursive search from the beginning of the string
        return ans; // Return the vector containing all valid word break combinations
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        vector<string> dict;
        string s;
        cin >> n;
        for (int i = 0;i < n;i++) {
            cin >> s;
            dict.push_back(s);
        }
        cin >> s;

        Solution ob;
        vector<string> ans = ob.wordBreak(n, dict, s);
        if (ans.size() == 0)
            cout << "Empty\n";
        else {
            sort(ans.begin(), ans.end());
            for (int i = 0;i < ans.size();i++)
                cout << "(" << ans[i] << ")";
            cout << endl;
        }
    }
    return 0;
}
// } Driver Code Ends

File: Python/BackTracking/03_word_break.py
"""
Idea:
using loop to find word which is in dictionary & recursively calling itself for next words till all words match the dictionary
"""


# User function Template for python3
class Solution:
    def wordBreak(self, n, dict, s):
        # Function to find all possible word break combinations

        def solve(arr, st, l, res):

            # If the entire string has been processed, append the current combination to the result
            if st == len(s):
                res.append(l[:])

            # Try all possible words from the dictionary and check if they match the substring
            for word in arr:
                if s[st:st + len(word)] == word:
                    l.append(word)  # Add the word to the current combination
                    solve(arr, st + len(word), l, res)  # Recursively check for the remaining substring
                    l.pop()  # Remove the last word from the current combination to backtrack

        # Initialize an empty list to store the result
        res = []

        # Call the recursive function to find all possible word break combinations
        solve(dict, 0, [], res)

        # Convert the combinations to string format
        res = [" ".join(x) for x in res]

        # Return the final result
        return res


# code here


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == '__main__':
    t = int(input())
    for _ in range(t):
        n = int(input())
        dicti = input().split()
        s = input()

        ob = Solution()
        ans = ob.wordBreak(n, dicti, s)
        if (len(ans) == 0):
            print("Empty")
        else:
            ans.sort()
            for it in (ans):
                print("(" + it, end=")")
            print()
# } Driver Code Ends